MOAPT-2671 - Update Jira plugin for the new Atlassian board view.

Alex-McEvoy opened the pull request:
### Jira URL
[cds-global.atlassian.net/browse/MOAPT-2671](https://cds-global.atlassian.net/browse/MOAPT-2671)

### Description
Atlassian is forcing all users of Jira to use the new look on Nov 5th.

Update the Jira plugin that shows PR labels for the new Atlassian view, which changes the page structure significantly.
The new look removes card elements from the DOM as they are scrolled out of view, and re-renders them when they are back in view. This is a shift from the previous way that cards were rendered, and requires us to listen for new cards creation.

1. Update css selectors for pertinent elements. 
2. Use MutationObservers to listen for card creation when scrolling
3. Cache previously observed card info, to speed up label insertion.

### QA Instructions
I'm not sure if there's a better way of doing this. Let me know if you know of one.
Get a zip file from Alex, and unzip it somewhere on your computer. You'll use that in step 10.

1. Make sure you're in the new look in Jira 
<img width="1445" alt="Screen Shot 2023-10-23 at 11 53 48 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/9b522744-87db-481c-b2b5-1a505a90fb53">


2. unzip the downloaded zip file sent to you

3. Go to the chrome extensions manager 

<img width="1369" alt="Screen Shot 2023-10-23 at 11 41 29 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/23928a6f-8c9c-4815-a6d2-0467b1719286">



4. Turn off the existing extension `Github Labels for JIRA Sprints 0.14.0`

<img width="1220" alt="Screen Shot 2023-10-23 at 11 41 56 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/14a79089-fa31-4cfa-88f5-da75742db905">



5. click on the "Load Unpacked" button 

<img width="785" alt="Screen Shot 2023-10-23 at 11 43 09 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/dd6f52a3-57af-4564-8995-147d7bddfb84">



6. Select the unzipped folder that you unzipped earlier, the whole folder, with everything in it should be called something like `gh-labels-in-jira`

<img width="810" alt="Screen Shot 2023-10-23 at 11 44 47 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/ae2f34aa-3b25-435d-9efa-c2e8560819fa">



7. Be sure the extension is turned on, and the version is `0.14.1`

<img width="798" alt="Screen Shot 2023-10-23 at 11 46 09 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/2bc1386e-56c6-4034-9417-3d5f26a1a9af">



8. Go to the sprint board (or any sprint board) https://cds-global.atlassian.net/jira/software/c/projects/MOAPT/boards/275, and see that the extension is there, you might need to click on the extension icon, and pin the plugin

<img width="1353" alt="Screen Shot 2023-10-23 at 11 47 27 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/2fa17d92-b074-4390-aa12-db1f0ce76d71">


9. click on the green icon in the toolbar, and select "options"

<img width="1495" alt="Screen Shot 2023-10-23 at 11 49 07 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/37a05a12-3721-40b0-8c07-957014a7ba99">



10. you'll need a github access token for this part, there's instructions on how to generate a new one here. https://www.notion.so/mediaos/Github-Labels-in-Jira-b05f734721b8463f9b79adab7a2ba462?pvs=4#d41996adfbeb471e828548e74e7268db
After you're done, put that token in the field in the plugin here. It's recommended to keep this token in a secure place, because you might need it again if you install/uninstall the plugin more than once, and you won't be able to get it from github after you copy it initially. 

<img width="1346" alt="Screen Shot 2023-10-23 at 11 47 59 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/68ee35cb-d2f7-4d48-858b-299cacaf1320">


11. Go back to the sprint board, and refresh the page, you should see the labels being applied

<img width="1103" alt="Screen Shot 2023-10-23 at 11 52 27 AM" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17329467/dd8b3b8c-3d23-45ce-a35d-2d2b27aa7de0">


12. Be sure to check the teams kanban board as well. https://cds-global.atlassian.net/jira/software/c/projects/MOAPT/boards/276 Make sure you've added a column name from the kanban board in the "options" of the ticket, an example might be "review needed". (it's not case sensitive anymore). You can also select "Show PRs in all columns" to see all PRs associated with all tickets.


13. Be sure to go back to the "options" by clicking on the plugin, and test the functionality of all the various buttons there. `Hide labels on closed PR's`, `Hide closed PR's` etc.



### Acceptance Criteria
Given that I navigate to a Jira view for a given board, when I look a the tickets that have a pull request associated, then I would see the Github labels for that ticket displayed below each ticket.

Alex-McEvoy committed:
Fix: Update for new Jira view

Alex-McEvoy committed:
Remove unnecessary asset exposure

Alex-McEvoy committed:
Feat: support kanban boards and select all columns

Alex-McEvoy committed:
Remove .ds_store

Alex-McEvoy committed:
Fix typo

Alex-McEvoy committed:
Fix refresh button, optimize to not run on backlog/timeline

sondoselayyanOP commented on content.js:
```diff
@@ -300,39 +275,106 @@ function addCodeReviewers(owner, repo, prid, label_id, requested_reviewers, labe
+            loadAwesomely(cardNode);
+        }
+        
+        if (MEMOIZED_CARDS[cardNode]) {
```
```suggestion
        if (MEMOIZED_CARDS[cardNode?.id]) {
```

sondoselayyanOP replied:
Question: should use the id of cardNode instead of cardNode because we are setting the cardId as the key in MEMOIZED_CARDS?

Alex-McEvoy replied:
ahh yes, good catch! Updated

Alex-McEvoy committed:
Fix: Update cached card retrieval

sondoselayyanOP commented:
@Alex-McEvoy I tried to active the `Hide labels on closed PR's` and `Hide closed PR's` , but they didn't work

jana-998 commented:
@Alex-McEvoy  I am having the same problem as Sondos, for `Hide labels on closed PR's` and `Hide closed PR's` 
Other wise LGTM!

Alex-McEvoy committed:
Feat: add hide merged PRs

Alex-McEvoy commented:
@sondoselayyanOP @jana-998 Hey thanks for taking a look all! Good catch on the hiding PRs feature.

So it looked like before, the `Hide closed PR's` and the `Hide labels on closed PR's` buttons did essentially the same thing, just hid "DECLINED" PRs. Hiding labels is the same thing as hiding PRs. Note that closed PRs here didn't mean "MERGED" PRs.

I've added another button to hide "MERGED" PRs, and removed the redundant `Hide labels on closed PR's`. The updates should hide both "DECLINED" and "MERGED" PRs, according to the user selection. 

It might be helpful to either add the "done" column to your watched columns, or just select the "Show PRs in all columns" option, to see the changes in the done column. 

Does that solve the issue you were seeing? Let me know if not, thanks again.

sondoselayyanOP reviewed (approved):
LGTM!

Alex-McEvoy committed:
Update loading

Alex-McEvoy committed:
Remove unneeded image

TomMokrzecki commented:
Works! Love my labels, ty!
QA approved, marked ready for FR

hm-ysoliz commented:
URLs tested:
https://cds-global.atlassian.net/jira/software/c/projects/COLI/boards/391
https://cds-global.atlassian.net/jira/software/c/projects/COLI/boards/256
https://cds-global.atlassian.net/jira/software/c/projects/COTRO/boards/257
https://cds-global.atlassian.net/jira/software/c/projects/MOAPT/boards/275

Steps:

- Downloaded and unzipped file.
- Turned off existing Chrome extension (version 0.14)
- Turned on Enhanced version of JIRA
- Loaded unpacked extension into Chrome (version 0.14.1)
- Generated a new token using the instructions here (https://www.notion.so/mediaos/Github-Labels-in-Jira-b05f734721b8463f9b79adab7a2ba462) including SSO auth
- Updated extension within Options with new token and tested various iterations (hiding merged PRs, hiding closed PRs, Code Reviewers, and showing PRs in all columns)

Feature Review Passed ✅ 

<img width="425" alt="image" src="https://github.com/Media-Platforms/gh-labels-in-jira/assets/17183938/b953e158-76c6-42c7-bdc6-7488a5cd0c85">

hm-ysoliz reviewed (approved):
Feature Review Passed (in comments)
